### DISCUSSION TITLE
Custom serializer in pretty printer for binary fields

### DISCUSSION THREAD

**Discussion by martinpaljak** [FasterXML/jackson-databind#3940] [Q&A] [answered]
Hi! I'd like the pretty printer to print binary as hex instead of base64. I've created a mapper and custom serializer for byte[].class that works in other contexts, but not in the pretty printer derived from that mapper, which still emits base64. 

Any hints?

**Comment by cowtowncoder** [comment]
Pretty printer really should not change serialization aspects; it should invoke all the usual handling so this sounds odd and not at all like expected.
If you could provide a reproduction that'd be good maybe we could help with suggestions.
Filing a separate bug issue is fine too but if repro is small enough can just continue here for now.

**Comment by martinpaljak** [comment, marked as answer]
My assumption was incorrect. I had a bunch of serializers for special objects "BlahAsBytesSerializer" in a module (for CBORMapper), so by just adding "BinaryAsHex" serializer to the end of same module and using the module with ObjectMapper (used for printing) did not "stack". A member of type Blah was first converted to byte[] and then printed as base64 whereas a simple byte[] was indeed serialized as hex.

"fixed" this by making "BlahAsHex" serializer. Would be cool if the mapper respected such two stage conversions, first from custom object to byte[] and then because same module has a serializer for byte[], would use that?

**Reply by cowtowncoder** [reply]
Unfortunately there is no way to stack serializers since they take Java Objects and produce `JsonToken` streams.

There are ways to do 2-phase conversions (using `Converter` with `JsonSerializer`) to allow Java Ob to Java Object conversions (on ser or deser) but that's about it; but one cannot dynamically mix and match them; need specific `JsonSerializer` that applies `Converter`. This is easy enough to do, but I don't think that'd do what you want here.

**Reply by martinpaljak** [reply]
Hijacking the same thread for a similar question - things work when pretty-printing a POJO with byte[] fields but reading a generic JsonNode from a stream and sending this to the same pretty printer will fail (read: results in base64). I thought that maybe adding a serializer for BinaryNode would fix it, but this does not get triggered.

How can I make any BinaryNode be pretty-printed in hex with a pretty printer bound to an object mapper?

**Reply by cowtowncoder** [reply]
@martinpaljak Challenge with `BinaryNode` serializer is that existence of custom serializer might not be checked properly yet -- for performance reasons delegation is likely used directly as `JsonNode` implementations implement `JsonSerializable`. I think override does work for `ArrayNode` and `ObjectNode` but not other types.
So you may want to file an issue here to make custom serializer override work (along with simple reproduction); it may be easy enough to add support.

Alternatively what is available now.... you could use wrapper type `POJONode` for values that implement `JsonSerializable` -- if so, matching method will get called. You can see

```
    public final void serialize(JsonGenerator gen, SerializerProvider ctxt) throws IOException
```

of `POJONode` for details. Actually, looks like you might be able to use `RawValue` as thing to pass: `JsonNodeFactory` has:

```
    @Override
    public ValueNode rawValueNode(RawValue value) {
        return new POJONode(value);
    }
```

which basically supports, I think, "injection" of `RawValue`s where you can pre-format output (but ensuring to use surrounding double-quotes; it's literally embedded in output).
